File: dart/lib/src/twitter/get_twitter_registry.dart
import '../rpc/rpc_client.dart';
import 'get_twitter_registry_key.dart';

/// Get Twitter registry state for a Twitter handle
///
/// This function mirrors js/src/twitter/getTwitterRegistry.ts
///
/// [connection] - The RPC client for Solana blockchain communication
/// [twitterHandle] - The Twitter handle (without @)
///
/// Returns the registry account info
Future<AccountInfo> getTwitterRegistry(
  RpcClient connection,
  String twitterHandle,
) async {
  final twitterHandleRegistryKey = await getTwitterRegistryKey(twitterHandle);

  final accountInfo = await connection.fetchEncodedAccount(
    twitterHandleRegistryKey.toBase58(),
  );

  if (!accountInfo.exists) {
    throw Exception(
        'Twitter registry does not exist for handle: $twitterHandle');
  }

  return accountInfo;
}

File: js/src/twitter/getTwitterRegistry.ts
import { Connection } from "@solana/web3.js";
import { TWITTER_ROOT_PARENT_REGISTRY_KEY } from "../constants";
import { NameRegistryState } from "../state";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Retrieves the user-facing name registry for a Twitter handle.
 *
 * @param connection Solana RPC connection
 * @param twitter_handle Twitter handle without the `@` prefix
 * @returns Name registry associated with the Twitter handle
 *
 * @example
 * ```ts
 * const registry = await getTwitterRegistry(connection, "bonfida");
 * ```
 */
export async function getTwitterRegistry(
  connection: Connection,
  twitter_handle: string,
): Promise<NameRegistryState> {
  const hashedTwitterHandle = getHashedNameSync(twitter_handle);
  const twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );
  const { registry } = await NameRegistryState.retrieve(
    connection,
    twitterHandleRegistryKey,
  );
  return registry;
}
